### PR TITLE
Keith/upgrade to new standards

### PR DESCRIPTION
**Pull request by keitheous**
[OFFAPPS-1073] Refactor app to meet new standards

## Description
Upgrade to new standards and look at ways to optimize code if have time.

## Tasks
- [x] Rosetta Path to en.yml - make a pr there too
- [ ] Update deployment scripts in samson ~and delilah~ when deploying
- [x] New standards - use webpack and package.json
- [x] New Zendesk Garden
- [x] Write one test - add more later
- [x] Implement Linter
- [x] Modularize main.js into app.js and ticket_sidebar.js (import export)
- [x] `vars` -> `lets` and `consts`
- [x] Apart from all the above, look at ways to optimize code

## Implementation notes
<details>
<summary>Removing Option to open url link in current tab as V2 uses Iframe</summary>
<em>Before (First) & After (Second) </em>
screen shot 2018-10-18 at 4 52 01 pm
screen shot 2018-10-18 at 4 51 32 pm
<br />
<em> The link will now always open in a new tab (by default).</em>
</details>
<br />
<details>
<summary>Optimizing Code `#convertPlaceholder` - Before (First) & After (Second)</summary>
before op
after op
<em>We can avoid the GET request if there is no string matches PLUS avoid assigning a variable if we were returning anyway. </em>
</details>
<br />

<details>
<summary>Separating events, app and ticket_sidebar - Before (First) & After (Second)</summary>
<br />
<em>assets > main.js</em>
<br />
screen shot 2018-10-17 at 10 32 31 am
<br />
<br />
becomes
<br />
<br />
<em>src > javascript > ticket_sidebar.js</em>
<br />
screen shot 2018-10-17 at 10 27 14 am
<br />
<em>src > javascript > events.js</em>
<br />
screen shot 2018-10-17 at 10 27 31 am
<br />
<em>src > javascript > app.js</em>
<br />
screen shot 2018-10-17 at 10 28 30 am
</details>

<br />
<details>
<summary>Use the same get request for ticket and ticket.assignee - Before (First) & After (Second)</summary>
screen shot 2018-10-18 at 3 01 52 pm
<br />
screen shot 2018-10-18 at 3 01 16 pm

<em> I wonder if there is a good reason why they did this in the past. Perhaps it is a V1 thing. In V2, we would already have the `assignee` object within the `ticket` object. In other words, `ticket.assignee` is available to us as it is located within `ticket`</em>
</details>

## References
[JIRA ticket](https://zendesk.atlassian.net/browse/OFFAPPS-1073)
[Rosetta Update Localization PR](https://github.com/zendesk/rosetta/pull/2381)

## CCs
@zendesk/apps-migration

## Risks
* There is a risk when converting `var`s to either `let`s or `const`s. If a data type is meant to be overwritten, using `const` instead of `let` would be a problem because const cannot be overwritten.
* In development, settings is loaded into app via app.registered returned data. In production, settings is configurable upon installation. I am confident my solution would work, please refer to `ticket_sidebar.js`, but I need some confirmation on this.
* Linter (standard) wants me to modify this a little. Should be okay. I removed some escape characters.

<details>
<summary>Before (First) & After (Second)</summary>
Before
After
</details>

### CODE REVIEW

File: spec/app.spec.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/* eslint-env jest */
+import app from '../src/javascript/app'
+
+import mockClient from '../src/javascript/zaf_client'
+jest.mock('../src/javascript/zaf_client')
+mockClient.get = jest.fn().mockResolvedValue(
```
Comment (keitheous) [inline spec/app.spec.js]: thanks @yu-tian113

File: src/javascript/app.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    // Thanks to github:bryanwoods for the pattern
+    const urlPattern = /(^|\s)(\b(https?):\/\/[A-Z0-9+&@#%?=~_|!:,.;]*[A-Z0-9+&@#%=~_|]\b)/ig
+    return string.replace(urlPattern, "$1<a href='$2' target='_blank'>$2</a>")
+  },
+
+  convertLinebreaks: function (string) {
+    return string.replace(/\n\r?/g, '<br />')
+  },
+
+  convertPlaceholders: function (string) {
```
Comment (yu-tian113) [inline src/javascript/app.js]: This function can return String or Promise, it's hard for the caller to know which will be returned.
What do you think of wrapping string in a promise?

File: src/javascript/app.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import client from './zaf_client'
+
+const app = {
+  renderText: function (settings) {
+    app.convertPlaceholders(settings.textBody)
```
Comment (yu-tian113) [inline src/javascript/app.js]: `convertPlaceholders` can return String, `string.then` will cause error

File: src/javascript/app.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      return {
+        ticket: data.ticket,
+        assignee: data.ticket.assignee,
+        currentUser: data.currentUser
+      }
+    })
+  },
+
+  resize: function () {
+    setTimeout(function () {
```
Comment (yu-tian113) [inline src/javascript/app.js]: need timeout?

Comment (keitheous) [reply src/javascript/app.js]: This part was brought over from V1 App, therefore I left this in there. Do you think we should take it out?

File: src/javascript/app.js (hunk 2 above)
Comment (keitheous) [reply src/javascript/app.js]: Thank you for noticing. Phew. How do you feel about this

`if (!strMatches) return Promise.resolve(string);`

File: src/javascript/app.js (hunk 3 above)
Comment (keitheous) [reply src/javascript/app.js]: Owh yeah! Thankfully you spot this

File: package.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  "devDependencies": {
+    "babel-jest": "^23.6.0",
+    "jest": "^23.6.0",
+    "standard": "^12.0.1"
+  },
+  "dependencies": {
+    "babel-core": "^6.26.3",
+    "babel-loader": "^7.1.4",
+    "babel-polyfill": "^6.26.0",
```
[the remark this answers is not in the document]
Comment (keitheous) [reply package.json]: Yes I will remove this

File: src/javascript/app.js (hunk 2 above)
Comment (keitheous) [reply src/javascript/app.js]: I've gone ahead and implemented `if (!strMatches) return Promise.resolve(string)`. Let me know if you have another suggestion =)

Comment (keitheous) [reply src/javascript/app.js]: tony gave me a thumb IRL

File: src/javascript/ticket_sidebar.js [hunk 6]
```diff
@@ -1,140 +1,134 @@
...
+  convertPlaceholders: function (string) {
+    const regexPlaceholder = new RegExp('({{)(.*?)(}})', 'g')
+    let strMatches = string.match(regexPlaceholder)
+    if (!strMatches) return Promise.resolve(string)
 
-    return app.getData().then(function(data) {
-      if (!strMatches) return string;
+    return app.getData().then(function (data) {
+      strMatches.forEach(function (curMatch) {
+        const regexSymbol = new RegExp('({{|}})', 'g')
```
Comment (ocke) [inline src/javascript/ticket_sidebar.js]: you can move this const out of the loop.

**Comment by keitheous** [comment]
Two thumbs up wohoo! Merging!

**Pull request merged**